registrations/serializers.py

```diff
@@ -273,15 +238,118 @@ def validate(self, data):
 
 
 class SeatReservationCodeSerializer(serializers.ModelSerializer):
-    timestamp = DateTimeField(default_timezone=pytz.UTC, required=False)
+    seats = serializers.IntegerField(required=True)
+
+    timestamp = DateTimeField(default_timezone=pytz.UTC, required=False, read_only=True)
+
     expiration = serializers.SerializerMethodField()
 
-    class Meta:
-        fields = ("seats", "code", "timestamp", "registration", "expiration")
-        model = SeatReservationCode
+    in_waitlist = serializers.SerializerMethodField()
 
     def get_expiration(self, obj):
-        return obj.timestamp + timedelta(minutes=code_validity_duration(obj.seats))
+        return obj.expiration
+
+    def get_in_waitlist(self, obj):
+        registration = obj.registration
+        maximum_attendee_capacity = registration.maximum_attendee_capacity
+
+        if maximum_attendee_capacity is None:
+            return False
+
+        attendee_count = registration.signups.filter(
+            attendee_status=SignUp.AttendeeStatus.ATTENDING
+        ).count()
+
+        return maximum_attendee_capacity - attendee_count <= 0
+
+    def to_internal_value(self, data):
+        if self.instance:
+            data["registration"] = self.instance.registration.id
+        return super().to_internal_value(data)
```

Comment:
`to_intermal_value` is responsible for deserialization. So in that sense this would be better done elsewhere.

I think `__init__` might work the best.

```
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        if self.instance:
            self.fields["registration"].read_only = True
```

Reply:
Personally, I wouldn't touch `__init__` in cases like these 😅 I think `to_internal_value` makes sense exactly because it's responsible for deserialization (i.e. write operations) *and* validation. It's the method that's used to return the values needed for building the instance.

Reply:
And FWIW it was something I suggested: https://github.com/City-of-Helsinki/linkedevents/pull/600#discussion_r1213055315

Reply:
Other options could be:
Do it in `validate_registration` method as suggested in https://github.com/City-of-Helsinki/linkedevents/pull/600#discussion_r1213088698
e.g.
```
def validate_registration(self, value):
        instance = self.instance
        if instance and instance.registration != value:
            raise serializers.ValidationError(
                {
                    _( "You may not change the registration of an existing object.")
                }
            )
        return value
```

or do in in update method as similar validation is done in SignUpSerializer and KeywordSetSerializer
e.g. 
```
def update(self, instance, validated_data):
        if (
            "registration" in validated_data
            and instance.registration != validated_data["registration"]
        ):
            raise serializers.ValidationError(
                {
                    "registration": _(
                        "You may not change the registration of an existing object."
                    )
                }
            )
            ...
```

Reply:
My reasoning for doing it in the `__init__` was that `self.fields["registration"].read_only = True` is almost self documenting way to do it. I also like `validate_registration` since there you can easily raise the validation error and API user is not left wondering why his update is not updating the value. :shrug:

Reply:
I can accept `to_internal_value` if a comment is added.

Reply:
Now that I think of it, the `__init__` option wouldn't work; setting the field to read-only during init would make the serializer skip the field completely during an update, i.e. it's not included in the data.

Reply:
Nope, the `to_internal_value` implementation doesn't skip the field, it just replaces it and also ensures that the data is there, regardless of what happens. Here's what happens if I remove the `to_internal_value` implementation and use the `__init__` implementation instead:
```
===================================================================================== short test summary info ======================================================================================
FAILED registrations/tests/test_seatsreservation_put.py::test_update_seats_reservation - KeyError: 'registration'
FAILED registrations/tests/test_seatsreservation_put.py::test_seats_amount_has_not_limit_if_maximum_attendee_capacity_is_none - KeyError: 'registration'
FAILED registrations/tests/test_seatsreservation_put.py::test_cannot_update_registration - KeyError: 'registration'
FAILED registrations/tests/test_seatsreservation_put.py::test_cannot_update_expired_reservation - KeyError: 'registration'
FAILED registrations/tests/test_seatsreservation_put.py::test_cannot_update_timestamp - KeyError: 'registration'
FAILED registrations/tests/test_seatsreservation_put.py::test_cannot_reserve_seats_if_there_are_not_enough_seats_available - KeyError: 'registration'
FAILED registrations/tests/test_seatsreservation_put.py::test_update_seats_reservation_in_waiting_list - KeyError: 'registration'
FAILED registrations/tests/test_seatsreservation_put.py::test_waiting_list_seats_amount_has_not_limit_if_waiting_list_capacity_is_none - KeyError: 'registration'
FAILED registrations/tests/test_seatsreservation_put.py::test_cannot_reserve_seats_waiting_list_if_there_are_not_enough_seats_available - KeyError: 'registration'
```

Reply:
About the other points; you're correct that my `to_internal_value` suggestion doesn't really tell the API user why the registration doesn't get updated and I think it's a valid point. Personally, I prefer convenience over convention in this case; telling the user that you can't change this value doesn't really bring any value.